src/array/cuda/utils.h

```diff
@@ -46,7 +46,7 @@ int _NumberOfBits(const T& range) {
   }
 
   int bits = 1;
-  while (bits < static_cast<int>(sizeof(T) * 8) && (1 << bits) < range) {
+  while (bits < static_cast<int>(sizeof(T) * 8) && (1ll << bits) < range) {
```

Comment:
what if `int64_t range = std::numeric_limits<int64_t>::max()`? and what if `T is unsigned`?

Reply:
range is of type T, so if T is unsigned, range can't be int64_t max.

Reply:
![image](https://github.com/dmlc/dgl/assets/85214957/1cb8f7b4-3e4f-4da0-a301-bddcc510f361)

Reply:
Did you test with `range = std::numeric_limits<uint64_t>::max()`?

Reply:
I think the problem is that when we try to shift 64 times for the unsigned case, it is like no shifting. Thus, an if check is required to see if the bits == maximum bit width before the first assertion.

Reply:
See here: https://godbolt.org/z/T4fTEsc5K
Pushing with this fix. Thanks for the catch!

Reply:
In C/C++ [bitwise shifting](https://en.wikipedia.org/wiki/Logical_shift) a value by a number of bits which is either a negative number or is greater than or equal to the total number of bits in this value results in undefined behavior.
Source: https://en.wikipedia.org/wiki/Undefined_behavior#:~:text=In%20C%2FC%2B%2B%20bitwise%20shifting,value%20results%20in%20undefined%20behavior.